src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,314 @@
+                    {
+                        case '.':
+                            if (i > 1 && i == length - 1 && c[i - 1] == '*')
+                            {
+                                sb.Length--;
+                                sb.Append('<'); // DOS_STAR (ends in *.)
+                            }
+                            else if (i < length - 1 && (c[i + 1] == '?' || c[i + 1] == '*'))
+                            {
+                                sb.Append('\"'); // DOS_DOT
+                            }
+                            else
+                            {
+                                sb.Append('.');
+                            }
+                            break;
+                        case '?':
+                            sb.Append('>'); // DOS_QM
+                            break;
+                        default:
+                            sb.Append(c[i]);
+                            break;
+                    }
+                }
+            }
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        /// <param name="expression">The expression to match with, such as "*.foo".</param>
+        /// <param name="name">The name to check against the expression.</param>
+        /// <param name="ignoreCase">True to ignore case (default).</param>
+        /// <remarks>
+        /// This is based off of System.IO.PatternMatcher used in FileSystemWatcher, which is based off
+        /// of RtlIsNameInExpression, which defines the rules for matching DOS wildcards ('*', '?', '&lt;', '&gt;', '"').
+        /// 
+        /// Like PatternMatcher, matching will not line up with Win32 behavior unless you transform the expression
+        /// using <see cref="TranslateExpression(string)"/>
+        /// </remarks>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
+        {
+            string debug = new string(name);
+
+            // The idea behind the algorithm is pretty simple. We keep track of all possible locations
+            // in the regular expression that are matching the name. When the name has been exhausted,
+            // if one of the locations in the expression is also just exhausted, the name is in the
+            // language defined by the regular expression.
+
+            if (string.IsNullOrEmpty(expression) || name.Length == 0)
+                return false;
+
+            if (expression[0] == '*')
+            {
+                // Just * matches everything
+                if (expression.Length == 1)
+                    return true;
+
+                if (expression.IndexOfAny(s_wildcardChars, startIndex: 1) == -1)
+                {
+                    // Handle the special case of a single  starting *, which essentially means "ends with"
```

Comment:
Nit: extra space between single and starting

Reply:
Thanks